Cython/Compiler/Parsing.py

```diff
@@ -2064,10 +2085,22 @@ def p_try_statement(s):
     else:
         s.error("Expected 'except' or 'finally'")
 
+
 def p_except_clause(s):
+    return _p_except_or_except_star_clause(s, False)
+
+def p_except_star_clause(s):
+    return _p_except_or_except_star_clause(s, True)
+
+def _p_except_or_except_star_clause(s, is_except_star):
+    # Share as much implementation as possible between except and except*
     # s.sy == 'except'
     pos = s.position()
     s.next()
+    if is_except_star:
+        s.expect('*')
+    elif s.sy == '*':
+        s.error("cannot have both 'except' and 'except*' on the same 'try'")
```

Comment:
I'm surprised that the `ExceptClauseNode` does not know whether it's `except` or `except*`. That seems to be the reason why you need to know up-front whether we're parsing one or the other. Wouldn't it be enough to just parse the clause and then ask it whether it's `except` or `except*`?

Reply:
I've changed this to be stored on `ExceptClauseNode` and with "cannot have both except and except*" enforced afterwards. I agree this is simpler.

I think the original version was an overly-literal translation of the Python grammar (where except and except* are considered different and unrelated to each other)